File: Application/src/main/java/com/shareyourproxy/app/fragment/SearchFragment.java
package com.shareyourproxy.app.fragment;

import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.text.Editable;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.shareyourproxy.IntentLauncher;
import com.shareyourproxy.R;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.JustObserver;
import com.shareyourproxy.api.rx.RxHelper;
import com.shareyourproxy.api.rx.RxQuery;
import com.shareyourproxy.api.rx.RxTextWatcherSubject;
import com.shareyourproxy.api.rx.event.OnBackPressedEvent;
import com.shareyourproxy.api.rx.event.RecyclerViewDatasetChangedEvent;
import com.shareyourproxy.api.rx.event.TextViewEditorActionEvent;
import com.shareyourproxy.api.rx.event.UserSelectedEvent;
import com.shareyourproxy.app.SearchActivity;
import com.shareyourproxy.app.adapter.BaseRecyclerView;
import com.shareyourproxy.app.adapter.SearchUserAdapter;
import com.shareyourproxy.app.adapter.SearchUserAdapter.UserViewHolder;
import com.shareyourproxy.widget.ContentDescriptionDrawable;
import com.shareyourproxy.widget.CustomEditText;

import java.util.HashMap;

import butterknife.Bind;
import butterknife.BindColor;
import butterknife.BindDimen;
import butterknife.BindInt;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import rx.Observer;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

import static android.graphics.PorterDuff.Mode.SRC_ATOP;
import static android.support.v4.graphics.drawable.DrawableCompat.setTintList;
import static android.support.v4.graphics.drawable.DrawableCompat.setTintMode;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.shareyourproxy.IntentLauncher.launchUserProfileActivity;
import static com.shareyourproxy.app.adapter.BaseRecyclerView.ViewState.EMPTY;
import static com.shareyourproxy.app.adapter.BaseRecyclerView.ViewState.MAIN;
import static com.shareyourproxy.app.adapter.BaseViewHolder.ItemClickListener;
import static com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard;
import static com.shareyourproxy.util.ViewUtils.showSoftwareKeyboard;
import static com.shareyourproxy.util.ViewUtils.svgToBitmapDrawable;

/**
 * Search for {@link User}s.
 */
public class SearchFragment extends BaseFragment implements ItemClickListener {

    public RxQuery rxQuery = RxQuery.INSTANCE;
    @Bind(R.id.fragment_search_bar_container)
    LinearLayout searchBarContainer;
    @Bind(R.id.fragment_search_back_button)
    ImageView imageViewBackButton;
    @Bind(R.id.fragment_search_edittext)
    CustomEditText editText;
    @Bind(R.id.fragment_search_clear_button)
    ImageView imageViewClearButton;
    @Bind(R.id.fragment_search_recyclerview)
    BaseRecyclerView recyclerView;
    @Bind(R.id.fragment_search_empty_textview)
    TextView emptyTextView;
    @Bind(R.id.fragment_search_empty_view_container)
    LinearLayout emptyViewContainer;
    @Bind(R.id.fragment_search_loadingview)
    ProgressBar loadingView;
    @BindInt(android.R.integer.config_shortAnimTime)
    int _animationDuration;
    @BindDimen(R.dimen.common_svg_null_screen_mini)
    int sexBotSize;
    @BindDimen(R.dimen.common_svg_large)
    int dimenSvgLarge;
    @BindColor(R.color.common_gray)
    int colorGray;
    @BindColor(R.color.common_blue)
    ColorStateList colorBlue;
    private SearchUserAdapter _adapter;
    private CompositeSubscription _subscriptions;
    private RxTextWatcherSubject _textWatcherSubject = RxTextWatcherSubject.INSTANCE;
    private RxHelper rxHelper = RxHelper.INSTANCE;
    private boolean _needsUpdate = true;
    private int _userCount;

    /**
     * Constructor.
     */
    public SearchFragment() {
    }

    /**
     * Return a new instance for the parent {@link SearchActivity}.
     *
     * @return new {@link SearchFragment}
     */
    public static SearchFragment newInstance() {
        return new SearchFragment();
    }

    /**
     * Handle back button press in this fragments parent {@link SearchActivity}.
     */
    @OnClick(R.id.fragment_search_back_button)
    void onClickBack() {
        hideSoftwareKeyboard(editText);
        getActivity().onBackPressed();
    }

    @OnClick(R.id.fragment_search_empty_button)
    void onClickInviteFriend() {
        editText.setText("");
        IntentLauncher.INSTANCE.launchInviteFriendIntent(getActivity());
    }

    /**
     * Clear the search edit text and search for all users.
     */
    @OnClick(R.id.fragment_search_clear_button)
    void onClickClear() {
        editText.setText("");
        //TODO: Make this clear to the featured users.
        _adapter.clearUserList();
        loadingView.setVisibility(GONE);
        recyclerView.updateViewState(
            new RecyclerViewDatasetChangedEvent(_adapter, EMPTY));
    }

    /**
     * {@link android.text.TextWatcher#afterTextChanged(Editable)}
     *
     * @param editable new search string
     */
    @OnTextChanged(value = R.id.fragment_search_edittext,
        callback = OnTextChanged.Callback.AFTER_TEXT_CHANGED)
    void onSearchStringChanged(Editable editable) {
        _textWatcherSubject.post(editable.toString());
        loadingView.setVisibility(VISIBLE);
    }

    @Override
    public View onCreateView(
        LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_search, container, false);
        ButterKnife.bind(this, rootView);
        initialize();
        return rootView;
    }

    /**
     * Initialize this view.
     */
    private void initialize() {
        editText.setRxBus(getRxBus());
        imageViewBackButton.setImageDrawable(getBackArrowDrawable());
        imageViewClearButton.setImageDrawable(getClearSearchDrawable());
        initializeRecyclerView();
        tintProgressBar();
    }

    public void tintProgressBar() {
        setTintMode(loadingView.getIndeterminateDrawable(), SRC_ATOP);
        setTintList(loadingView.getIndeterminateDrawable(), colorBlue);
    }

    /**
     * Initialize a recyclerView with User data.
     */
    private void initializeRecyclerView() {
        emptyTextView.setCompoundDrawablesWithIntrinsicBounds(
            null, getSexBotDrawable(), null, null);
        _adapter = SearchUserAdapter.newInstance(recyclerView, this);

        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setEmptyView(emptyViewContainer);
        recyclerView.setHasFixedSize(true);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.addOnScrollListener(getDismissScrollListener());
        recyclerView.setAdapter(_adapter);
    }

    /**
     * Parse a svg and return a null screen sized {@link ContentDescriptionDrawable} .
     *
     * @return Drawable with a contentDescription
     */
    private Drawable getSexBotDrawable() {
        return svgToBitmapDrawable(getActivity(), R.raw.ic_sexbot, sexBotSize);
    }

    @Override
    public void onItemClick(View view, int position) {
        UserViewHolder holder = (UserViewHolder) recyclerView.getChildViewHolder(view);
        getRxBus().post(
            new UserSelectedEvent(
                holder.userImage, holder.userName, _adapter.getItemData(position)));
    }

    /**
     * Return an SVG image drawable icon for the back arrow.
     *
     * @return back arrow image.drawable
     */
    private Drawable getBackArrowDrawable() {
        return svgToBitmapDrawable(getActivity(), R.raw.ic_arrow_back, dimenSvgLarge, colorGray);
    }

    /**
     * Return an SVG image drawable icon for the clear button.
     *
     * @return clear button image.drawable
     */
    private Drawable getClearSearchDrawable() {
        return svgToBitmapDrawable(getActivity(), R.raw.ic_clear, dimenSvgLarge, colorGray);
    }

    @Override
    public void onResume() {
        super.onResume();
        _subscriptions = rxHelper.checkCompositeButton(_subscriptions);
        _subscriptions.add(getRxBus().toObservable()
            .subscribe(onNextEvent()));

        final User loggedInUser = getLoggedInUser();
        if (loggedInUser != null) {
            _subscriptions.add(
                _textWatcherSubject.toObserverable()
                    .compose(rxHelper.<String>observeMain())
                    .subscribe(getUsersObserver(loggedInUser)));
            //search entered text
            _textWatcherSubject.post(editText.getText().toString().trim());
        }
        showSoftwareKeyboard(editText);
    }

    public JustObserver<String> getUsersObserver(final User loggedInUser) {
        return new JustObserver<String>() {

            @Override
            public void next(String queryName) {
                String trimmedName = queryName.trim();
                if (!trimmedName.isEmpty()) {
                    _adapter.setQueryString(trimmedName);
                    rxQuery.searchMatchingUsers(getActivity(), trimmedName, loggedInUser.id())
                        .compose(rxHelper.<HashMap<String, User>>observeMain())
                        .subscribe(getSearchObserver());
                }
            }
        };
    }

    /**
     * Textwatcher Subject Observer. Refresh the user list based on a search query.
     *
     * @return search observer
     */
    private Observer<HashMap<String, User>> getSearchObserver() {
        return new Observer<HashMap<String, User>>() {
            @Override
            public void onCompleted() {
                _needsUpdate = true;
                if (_userCount == 0) {
                    recyclerView.updateViewState(
                        new RecyclerViewDatasetChangedEvent(_adapter, EMPTY));
                }
                loadingView.setVisibility(GONE);
            }

            @Override
            public void onError(Throwable e) {
                _needsUpdate = true;
                recyclerView.updateViewState(
                    new RecyclerViewDatasetChangedEvent(_adapter, EMPTY));
                loadingView.setVisibility(GONE);
                Timber.e("Error %1$s", Log.getStackTraceString(e));
            }

            @Override
            public void onNext(HashMap<String, User> users) {
                _userCount = users.size();
                if (_userCount > 0) {
                    if (_needsUpdate) {
                        _adapter.clearUserList();
                        _needsUpdate = false;
                    }
                    _adapter.refreshData(users.values());
                    recyclerView.updateViewState(
                        new RecyclerViewDatasetChangedEvent(_adapter, MAIN));
                    recyclerView.scrollToPosition(0);
                }
            }
        };
    }

    /**
     * Observe the next event.
     *
     * @return next event observer
     */
    private JustObserver<Object> onNextEvent() {
        return new JustObserver<Object>() {
            @Override
            public void next(Object event) {
                if (event instanceof UserSelectedEvent) {
                    onUserSelected((UserSelectedEvent) event);
                } else if (event instanceof OnBackPressedEvent) {
                    imageViewClearButton.animate().alpha(0f).setDuration(_animationDuration);
                } else if (event instanceof TextViewEditorActionEvent) {
                    if (((TextViewEditorActionEvent) event).getKeyEvent().getKeyCode() ==
                        KeyEvent.KEYCODE_DEL) {
                        if (editText.length() == 0) {
                            _adapter.clearUserList();
                            recyclerView.updateViewState(
                                new RecyclerViewDatasetChangedEvent(_adapter, EMPTY));
                            loadingView.setVisibility(GONE);
                        } else {
                            loadingView.setVisibility(VISIBLE);
                        }
                    }
                }
            }
        };
    }

    @Override
    public void onPause() {
        super.onPause();
        hideSoftwareKeyboard(getView());
        _subscriptions.unsubscribe();
        _subscriptions = null;
    }

    /**
     * User selected, open their profile.
     *
     * @param event data
     */
    public void onUserSelected(UserSelectedEvent event) {
        launchUserProfileActivity(getActivity(), event.getUser(), getLoggedInUser().id(),
            event.getImageView(), event.getTextView());
    }

}

File: Application/src/main/kotlin/com/shareyourproxy/app/fragment/SearchFragment.kt
package com.shareyourproxy.app.fragment

import android.R.integer.config_shortAnimTime
import android.content.res.ColorStateList
import android.graphics.PorterDuff.Mode.SRC_ATOP
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v4.graphics.drawable.DrawableCompat.setTintList
import android.support.v4.graphics.drawable.DrawableCompat.setTintMode
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.text.Editable
import android.text.TextWatcher
import android.util.Log.getStackTraceString
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.*
import com.shareyourproxy.IntentLauncher.launchInviteFriendIntent
import com.shareyourproxy.IntentLauncher.launchUserProfileActivity
import com.shareyourproxy.R
import com.shareyourproxy.R.color.common_blue
import com.shareyourproxy.R.color.common_gray
import com.shareyourproxy.R.dimen.common_svg_large
import com.shareyourproxy.R.dimen.common_svg_null_screen_mini
import com.shareyourproxy.R.layout.fragment_search
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.JustObserver
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.RxBusRelay.rxBusObservable
import com.shareyourproxy.api.rx.RxHelper.observeMain
import com.shareyourproxy.api.rx.RxQuery.searchMatchingUsers
import com.shareyourproxy.api.rx.RxTextWatcherRelay.post
import com.shareyourproxy.api.rx.RxTextWatcherRelay.textWatcherObserverable
import com.shareyourproxy.api.rx.event.OnBackPressedEvent
import com.shareyourproxy.api.rx.event.RecyclerViewDatasetChangedEvent
import com.shareyourproxy.api.rx.event.TextViewEditorActionEvent
import com.shareyourproxy.api.rx.event.UserSelectedEvent
import com.shareyourproxy.app.adapter.BaseRecyclerView
import com.shareyourproxy.app.adapter.BaseViewHolder.ItemClickListener
import com.shareyourproxy.app.adapter.SearchUserAdapter
import com.shareyourproxy.app.adapter.SearchUserAdapter.UserViewHolder
import com.shareyourproxy.util.ButterKnife.bindColor
import com.shareyourproxy.util.ButterKnife.bindColorStateList
import com.shareyourproxy.util.ButterKnife.bindDimen
import com.shareyourproxy.util.ButterKnife.bindInt
import com.shareyourproxy.util.ButterKnife.bindView
import com.shareyourproxy.util.Enumerations.ViewState.EMPTY
import com.shareyourproxy.util.Enumerations.ViewState.MAIN
import com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard
import com.shareyourproxy.util.ViewUtils.showSoftwareKeyboard
import com.shareyourproxy.util.ViewUtils.svgToBitmapDrawable
import com.shareyourproxy.widget.CustomEditText
import rx.Observer
import rx.subscriptions.CompositeSubscription
import timber.log.Timber
import java.util.*

/**
 * Search for [User]s.
 */
internal final class SearchFragment() : BaseFragment(), ItemClickListener {

    private val imageViewBackButton: ImageView by bindView(R.id.fragment_search_back_button)
    private val editText: CustomEditText by bindView(R.id.fragment_search_edittext)
    private val imageViewClearButton: ImageView by bindView(R.id.fragment_search_clear_button)
    private val recyclerView: BaseRecyclerView by bindView(R.id.fragment_search_recyclerview)
    private val emptyTextView: TextView by bindView(R.id.fragment_search_empty_textview)
    private val emptyViewContainer: LinearLayout by bindView(R.id.fragment_search_empty_view_container)
    private val emptyViewButton: Button by bindView(R.id.fragment_search_empty_button)
    private val loadingView: ProgressBar by bindView(R.id.fragment_search_loadingview)
    private val animationDuration: Int by bindInt(config_shortAnimTime)
    private val sexBotSize: Int by bindDimen(common_svg_null_screen_mini)
    private val dimenSvgLarge: Int by bindDimen(common_svg_large)
    private val colorGray: Int by bindColor(common_gray)
    private val colorBlue: ColorStateList by bindColorStateList(common_blue)
    private val adapter: SearchUserAdapter = SearchUserAdapter(recyclerView, this)
    private val subscriptions: CompositeSubscription = CompositeSubscription()
    private val sexBotDrawable: Drawable = svgToBitmapDrawable(activity, R.raw.ic_sexbot, sexBotSize)
    private val backArrowDrawable: Drawable = svgToBitmapDrawable(activity, R.raw.ic_arrow_back, dimenSvgLarge, colorGray)
    private val clearSearchDrawable: Drawable = svgToBitmapDrawable(activity, R.raw.ic_clear, dimenSvgLarge, colorGray)
    /**
     * Observe the next event.
     * @return next event observer
     */
    private val onNextEvent = object : JustObserver<Any>(SearchFragment::class.java) {
        @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
        override fun next(event: Any) {
            if (event is UserSelectedEvent) {
                onUserSelected(event)
            } else if (event is OnBackPressedEvent) {
                imageViewClearButton.animate().alpha(0f).setDuration(animationDuration.toLong())
            } else if (event is TextViewEditorActionEvent) {
                if (event.keyEvent.keyCode == KeyEvent.KEYCODE_DEL) {
                    if (editText.length() == 0) {
                        adapter.clearUserList()
                        recyclerView.updateViewState(RecyclerViewDatasetChangedEvent(adapter, EMPTY))
                        loadingView.visibility = GONE
                    } else {
                        loadingView.visibility = VISIBLE
                    }
                }
            }
        }
    }
    /**
     * Handle back button press in this fragments parent [SearchActivity].
     */
    private val onClickBack: View.OnClickListener = View.OnClickListener {
        hideSoftwareKeyboard(editText)
        activity.onBackPressed()
    }

    private val onClickInviteFriend: View.OnClickListener = View.OnClickListener {
        editText.setText("")
        launchInviteFriendIntent(activity)
    }

    /**
     * Clear the search edit text and search for all users.
     */
    private val onClickClear: View.OnClickListener = View.OnClickListener {
        editText.setText("")
        //TODO: Make this clear to the featured users.
        adapter.clearUserList()
        loadingView.visibility = GONE
        recyclerView.updateViewState(RecyclerViewDatasetChangedEvent(adapter, EMPTY))
    }

    /**
     * [android.text.TextWatcher.afterTextChanged]
     */
    private val onSearchStringChanged: TextWatcher = object : TextWatcher {
        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun afterTextChanged(editable: Editable) {
            post(editable.toString())
            loadingView.visibility = VISIBLE
        }

    }

    /**
     * Textwatcher Subject Observer. Refresh the user list based on a search query.
     * @return search observer
     */
    private val searchObserver: Observer<HashMap<String, User>> = object : Observer<HashMap<String, User>> {
        override fun onCompleted() {
            needsUpdate = true
            if (userCount == 0) {
                recyclerView.updateViewState(
                        RecyclerViewDatasetChangedEvent(adapter, EMPTY))
            }
            loadingView.visibility = GONE
        }

        override fun onError(e: Throwable) {
            needsUpdate = true
            recyclerView.updateViewState(RecyclerViewDatasetChangedEvent(adapter, EMPTY))
            loadingView.visibility = GONE
            Timber.e("Error ${getStackTraceString(e)}")
        }

        override fun onNext(users: HashMap<String, User>) {
            userCount = users.size
            if (userCount > 0) {
                if (needsUpdate) {
                    adapter.clearUserList()
                    needsUpdate = false
                }
                adapter.refreshData(users.values)
                recyclerView.updateViewState(
                        RecyclerViewDatasetChangedEvent(adapter, MAIN))
                recyclerView.scrollToPosition(0)
            }
        }
    }

    private val getUsersObserver = object : JustObserver<String>(SearchFragment::class.java) {
        @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
        override fun next(queryName: String) {
            val trimmedName: String = queryName.trim { it <= ' ' }
            if (!trimmedName.isEmpty()) {
                adapter.setQueryString(trimmedName)
                searchMatchingUsers(activity, trimmedName, loggedInUser.id).compose(observeMain<HashMap<String, User>>()).subscribe(searchObserver)
            }
        }
    }

    private var needsUpdate = true
    private var userCount: Int = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(fragment_search, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initialize()
    }

    override fun onResume() {
        super.onResume()
        subscriptions.add(rxBusObservable().subscribe(onNextEvent))
        subscriptions.add(textWatcherObserverable().compose(observeMain<String>()).subscribe(getUsersObserver))
        //search entered text
        post(editText.text.toString().trim { it <= ' ' })
        showSoftwareKeyboard(editText)
    }

    override fun onPause() {
        super.onPause()
        hideSoftwareKeyboard(view)
        subscriptions.unsubscribe()
    }

    override fun onItemClick(view: View, position: Int) {
        val holder = recyclerView.getChildViewHolder(view) as UserViewHolder
        post(UserSelectedEvent(holder.userImage, holder.userName, adapter.getItemData(position)))
    }

    /**
     * Initialize this view.
     */
    private fun initialize() {
        editText.addTextChangedListener(onSearchStringChanged)
        imageViewBackButton.setOnClickListener(onClickBack)
        emptyViewButton.setOnClickListener(onClickInviteFriend)
        imageViewClearButton.setOnClickListener(onClickClear)
        imageViewBackButton.setImageDrawable(backArrowDrawable)
        imageViewClearButton.setImageDrawable(clearSearchDrawable)
        initializeRecyclerView()
        tintProgressBar()
    }

    private fun tintProgressBar() {
        setTintMode(loadingView.indeterminateDrawable, SRC_ATOP)
        setTintList(loadingView.indeterminateDrawable, colorBlue)
    }

    /**
     * Initialize a recyclerView with User data.
     */
    private fun initializeRecyclerView() {
        emptyTextView.setCompoundDrawablesWithIntrinsicBounds(null, sexBotDrawable, null, null)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.setEmptyView(emptyViewContainer)
        recyclerView.setHasFixedSize(true)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.addOnScrollListener(dismissScrollListener)
        recyclerView.adapter = adapter
    }

    /**
     * User selected, open their profile.
     * @param event data
     */
    private fun onUserSelected(event: UserSelectedEvent) {
        launchUserProfileActivity(activity, event.user, loggedInUser.id, event.imageView, event.textView)
    }
}
